aiobungie/internal/iterators.py

```diff
@@ -0,0 +1,574 @@
+        """Returns the first item in the iterator.
+
+        Example
+        -------
+        >>> iterator = FlatIterator([3, 1, 6, 7])
+        >>> iterator.first()
+        3
+
+        Raises
+        ------
+        `StopIteration`
+            If no elements are left in the iterator.
+        """
+        return self.take(1).next()
+
+    def last(self) -> Item:
+        """Returns the last item in the iterator.
+
+        Example
+        -------
+        >>> iterator = FlatIterator([3, 1, 6, 7])
+        >>> iterator.last()
+        7
+
+        Raises
+        ------
+        `StopIteration`
+            If no elements are left in the iterator.
+        """
+        return self.collect()[::-1][0]
+
+    def count(self) -> int:
+        count = 0
+        for _ in self:
+            count += 1
+
+        return count
+
+    def union(self, other: FlatIterator[Item]) -> FlatIterator[Item]:
+        """Returns an iterator that yields all items from both iterators.
+
+        Example
+        -------
+        >>> iterator = FlatIterator([1, 2, 3])
+        >>> other = FlatIterator([4, 5, 6])
+        >>> async for item in iterator.union(other):
+                print(item)
+        # 1
+        # 2
+        # 3
+        # 4
+        # 5
+        # 6
+
+        Parameters
+        ----------
+        other: `FlatIterator[Item]`
+            The iterable to union with.
+
+        Raises
+        ------
+        `StopIteration`
+            If no elements are left in the iterator.
+        """
+        return FlatIterator(itertools.chain(self, other))
+
+    def for_each(self, func: collections.Callable[[Item], typing.Any]) -> None:
+        """Calls the function on each item in the iterator.
+
+        Example
+        -------
+        >>> iterator = FlatIterator([1, 2, 3])
+        >>> iterator.for_each(lambda item: print(item))
+        # 1
+        # 2
+        # 3
+
+        Parameters
+        ----------
+        func: `typeshed.Callable[[Item], None]`
+            The function to call on each item in the iterator.
+        """
+        for item in self:
+            func(item)
+
+    def _ok(self) -> typing.NoReturn:
+        raise StopIteration("No more items in the iterator.")
+
+    def __repr__(self) -> str:
+        return f'<{self.__class__.__name__}({", ".join([str(item) for item in self])})>'
+
+    def __len__(self) -> int:
+        return self.count()
+
+    # We support both sync and async iter.
+    def __iter__(self) -> FlatIterator[Item]:
+        return self
+
+    def __aiter__(self) -> FlatIterator[Item]:
+        return self
+
+    def __next__(self) -> Item:
+        try:
+            item = next(self._items)
+        except StopIteration:
+            self._ok()
+
+        return item
+
+    async def __anext__(self) -> Item:
+        try:
+            return self.__next__()
+        except StopIteration as e:
+            raise StopAsyncIteration from e
+
+
+def into_iter(
+    iterable: typing.Iterable[Item],
```

Comment:
Use collections instead of typing.
```suggestion
    iterable: collections.Iterable[Item],
```